### PR TITLE
Deserialize values in model type

### PR DESCRIPTION
**Pull request by bradleesand**
This is my suggestion for fixing #99. It could certainly be DRYer but I just wanted to put this out for comments in case there's something that I overlooked.

### CODE REVIEW

**Comment by jrochkind** [comment]
Thanks, can you explain in words what you've done here?  At first glance it looks like just copying and pasting the `cast` implementation into `deserialize`, but that must not be it, as that wouldn't do anything. You copied and pasted, but changed something? Can you explain what?

Also, if we chose to go this way, we'd need a test that fails without the PR but passes with it, a test that ensures the implementation does what you need it to that it didn't do before.

File: spec/nested_attributes/nested_attributes_spec.rb [hunk 1]
```diff
@@ -369,4 +369,40 @@ def one_model_attributes=(attrs)
...
+      ActiveRecord::Type.register(:nested_serializing, serializing_type)
+      expect { instance.one_model }.to_not raise_error
+
+      instance.one_model = {str: 'foo'}
+      expect(instance.one_model.str).to eq 'foo'
+
+      instance.save!
+      instance.reload
+
+      expect(instance.one_model.str).to eq 'foo'
```
Comment (bradleesand) [inline spec/nested_attributes/nested_attributes_spec.rb:404]: Before the change to `deserialize`, this expectation fails:
```
  1) AttrJson::NestedAttributes model with serializing type supports custom ActiveRecord registered type that serializes value
     Failure/Error: expect(instance.one_model.str).to eq 'foo'
     
       expected: "foo"
            got: "foox"
     
       (compared using ==)
     # ./spec/nested_attributes/nested_attributes_spec.rb:404:in `block (3 levels) in <top (required)>'

Finished in 0.17165 seconds (files took 3.24 seconds to load)
1 example, 1 failure
```

File: lib/attr_json/model.rb [hunk 2]
```diff
@@ -104,15 +104,21 @@ def attr_json_config(new_values = {})
...
           # store keys in arguments get translated to attribute names on initialize.
-          if attribute_def = self.attr_json_registry.store_key_lookup("", key.to_s)
+          k = if attribute_def = self.attr_json_registry.store_key_lookup("", key.to_s)
             attribute_def.name.to_s
           else
             key
           end
-        end
+          v = if deserialize && (attr_def = model.attr_json_registry[k])
+            attr_def.deserialize(value)
```
Comment (bradleesand) [inline lib/attr_json/model.rb:116]: This is the core of this PR. We deserialize the values before calling `new`

**Comment by jrochkind** [comment]
Thanks! I'll try to look at this this soon!

**Comment by bradleesand** [comment]
Please let me know if you need more explanation of this change.

**Comment by jrochkind** [comment]
thank you! I'm sorry it's just taking me some time to get to looking at it, with some both personal and professional things keeping me busy. If you don't hear back in a week or two feel free to ping me again!

**Comment by bradleesand** [comment]
No worries! I totally understand. I did DRY it up a bit but I'm sure there's more that can be done.

**Comment by jrochkind** [comment]
Thanks for the DRY!

Sorry I've been neglecting this. I'm afraid I'm going to neglect it a bit more -- I have limited time for attr_json support, and am looking at refactoring some parts to stick more closely to ActiveRecord attributes -- I'm curious to see how/if my refactors effect the need for this code first. 

But not sure what my timeline is. 

I do understand your use case and would like to support it, just have limited bandwidth.

**Comment by jrochkind** [comment]
@bradleesand sorry for letting this sit so long, I'm trying to get to it!

How do you feel about giving me permission to commit to your branch? https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by jrochkind** [comment]
From original issue

> It works correctly when I use it on an ActiveRecord with attribute :password, :encrypted and even with attr_json :password, :encrypted

I have verified that it works with `attr_json :password` and a serializing type. But breaks when it's in a model. I am working to try to understand what's going on here better, why that would be any different.

**Comment by jrochkind** [comment]
aha, i'm getting the idea, right. hm.

**Comment by jrochkind** [comment]
OK, so investigating the code, I think I've run into some other existing behavior that I'd consider a bug; that fixing it can let us fix this with somewhat simpler code. 

An analogous feature we already have in the code base is the `store_key` -- you can say that an attribute should be stored in the json under a different name than the attribute name. 

```
class SomeModel < ApplicationRecord
   attr_json :str, :string, store_key: "string_key_in_db"
end
```

If you do `SomeModel.new(str: "foo")`, it'll end up in the json in the database as `{"json_attributes" = { "string_key_in_db" => "foo" }`, great. 

But what happens if you do `SomeMode.new(string_key_in_db: "foo")`? Well, you get a `ActiveModel::UnknownAttributeError`, great. That's the serialization key, you can't pass it in there. 

But what if you have a store_key in a nested model? Turns out it's not consistent. 

    record.new(nested_model: { string_key_in_db: "foo" })

works, and is *equivalent* to:

    record.new(nested_model: { string: "foo" })

I don't think this is actually right. There are no tests that insist upon it -- if I make this stop working, no tests fail. 

This is of course, about a similar confusion between casting and deserialization. 

If we fix/break/change this, then we can simplify the code in general while also fixing your problem. Because only deserialization should use that `new_from_serializable` at all (it really should be called `new_from_serialized`) -- casting should just use `new`. 

I am going to do that in another branch, to do things slightly differently, and cause I don't think I can commit to this branch. But I'll give you credit in CHANGELOG etc. 

Sorry this took so long @bradleesand, hope to have it settled soon, thanks for your interest and contribution!

**Comment by jrochkind** [comment]
Closing this for #125 instead. 

Sorry it took so long! @bradleesand , your preliminary work here to reproduce and suggest a solution helped a lot, thank you!

I may wait until next week to actually release a version with this fix. If you see this before then, and are still interested, and want to confirm this fixes your problem, that'd be great! But no worries, otherwise I'll release a new version with what I believe is a fix next week anyway.

**Pull request closed without merging**